lightning-c-bindings/src/c_types/mod.rs

```diff
@@ -0,0 +1,415 @@
+		Self { data: [0; 32] }
+	}
+}
+
+#[repr(C)]
+pub struct ThreeBytes { pub data: [u8; 3], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct FourBytes { pub data: [u8; 4], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct TenBytes { pub data: [u8; 10], }
+#[derive(Clone)]
+#[repr(C)]
+pub struct SixteenBytes { pub data: [u8; 16], }
+
+pub(crate) struct VecWriter(pub Vec<u8>);
+impl lightning::util::ser::Writer for VecWriter {
+	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
+		self.0.extend_from_slice(buf);
+		Ok(())
+	}
+	fn size_hint(&mut self, size: usize) {
+		self.0.reserve_exact(size);
+	}
+}
+pub(crate) fn serialize_obj<I: lightning::util::ser::Writeable>(i: &I) -> derived::CVec_u8Z {
+	let mut out = VecWriter(Vec::new());
+	i.write(&mut out).unwrap();
+	CVecTempl::from(out.0)
+}
+pub(crate) fn deserialize_obj<I: lightning::util::ser::Readable>(s: u8slice) -> Result<I, lightning::ln::msgs::DecodeError> {
+	I::read(&mut s.to_slice())
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+/// A Rust str object, ie a reference to a UTF8-valid string.
+/// This is *not* null-terminated so cannot be used directly as a C string!
+pub struct Str {
+	pub chars: *const u8,
+	pub len: usize
+}
+impl Into<Str> for &'static str {
+	fn into(self) -> Str {
+		Str { chars: self.as_ptr(), len: self.len() }
+	}
+}
+impl Into<&'static str> for Str {
+	fn into(self) -> &'static str {
+		if self.len == 0 { return ""; }
+		std::str::from_utf8(unsafe { std::slice::from_raw_parts(self.chars, self.len) }).unwrap()
+	}
+}
+
+// Note that the C++ headers memset(0) all the Templ types to avoid deallocation!
+// Thus, they must gracefully handle being completely null in _free.
+
+/// CSliceTempl may have the word slice in it, but its really a Vec. Because its used for things
+/// that require per-item conversion, we have to use a new buffer for it. Of course for opaque
+/// types, each item will simply be a pointer to the original slice's corresponding item, so we may
+/// not own the actual Rust item, only the buffer.
+///
+/// Thus, the implementation here is basically just that of CVecTempl.
+#[repr(C)]
+pub struct CSliceTempl<T> {
+	pub data: *mut T,
+	pub datalen: usize
+}
+impl<T> CSliceTempl<T> {
+	pub(crate) fn into_rust_vec(mut self) -> Vec<T> {
+		if self.datalen == 0 { return Vec::new(); }
+		let ret = unsafe { Box::from_raw(std::slice::from_raw_parts_mut(self.data, self.datalen)) }.into();
+		self.data = std::ptr::null_mut();
+		self.datalen = 0;
+		ret
+	}
+}
+impl<T> Drop for CSliceTempl<T> {
+	fn drop(&mut self) {
+		if self.datalen == 0 { return; }
+		unsafe { Box::from_raw(std::slice::from_raw_parts_mut(self.data, self.datalen)) };
+	}
+}
+
+// TODO: Integer/bool primitives should avoid the pointer indirection for underlying types
+// everywhere in the containers.
+
+#[repr(C)]
+pub union CResultPtr<O, E> {
+	pub result: *mut O,
+	pub err: *mut E,
+}
+#[repr(C)]
+pub struct CResultTempl<O, E> {
+	pub contents: CResultPtr<O, E>,
+	pub result_good: bool,
```

Comment:
FWIW you could use a rust enum here and it is represented the same in memory (and cbindgen should reflect that)

ie 
```rust 
#[repr(C, u8)]
enum TwoCases {
    A(u8, u16),
    B(u16),
}
```
Will be represented as:
```rust
#[repr(C)]
struct TwoCasesRepr {
    tag: TwoCasesTag,
    payload: TwoCasesPayload,
}
#[repr(u8)]
enum TwoCasesTag { A, B }

#[repr(C)]
union TwoCasesPayload {
   A: TwoCasesPayloadA,
   B: u16,
}

#[repr(C)]
struct TwoCasesPayloadA(u8, u16);
```

Reply:
Right, could do, though its a bit clearer to just use a bool instead of having cbindgen generate tags and all that mess.